{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8de3e634",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "!pip install -U matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49e87e4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import dxdata\n",
    "import dxpy\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyspark\n",
    "import toml\n",
    "from matplotlib.lines import Line2D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "574765c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../config.toml\") as f:\n",
    "    conf = toml.load(f)\n",
    "\n",
    "GENE = conf[\"GENE\"]\n",
    "# REFERENCE_GENOME = conf[\"REFERENCE_GENOME\"]\n",
    "# LOG_FILE = Path(conf[\"IMPORT\"][\"LOG_DIR\"], f\"{GENE}_results.log\").resolve().__str__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c613b451",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Spark and Hail\n",
    "sc = pyspark.SparkContext()\n",
    "spark = pyspark.sql.SparkSession(sc)\n",
    "\n",
    "# hl.init(sc=sc, default_reference=REFERENCE_GENOME, log=LOG_FILE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe74bd2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pval(x):\n",
    "    if x < 0.0001:\n",
    "        return \"****\"\n",
    "    if x < 0.001:\n",
    "        return \"***\"\n",
    "    if x < 0.01:\n",
    "        return \"**\"\n",
    "    if x < 0.05:\n",
    "        return \"*\"\n",
    "    if x > 0.05:\n",
    "        return \"\"\n",
    "\n",
    "\n",
    "def pheno_search(x):\n",
    "    try:\n",
    "        return ukb_coding[ukb_coding.FieldID == int(x)].Field.values[0]\n",
    "    except:\n",
    "        # coding from https://biobank.ndph.ox.ac.uk/showcase/coding.cgi?id=100605\n",
    "        if x == \"6150_1\":\n",
    "            return \"Heart attack diagnosed by doctor\"\n",
    "        elif x == \"6150_2\":\n",
    "            return \"Angina diagnosed by doctor\"\n",
    "        elif x == \"6150_3\":\n",
    "            return \"Stroke diagnosed by doctor\"\n",
    "        elif x == \"6150_4\":\n",
    "            return \"High blood pressure diagnosed by doctor\"\n",
    "        elif x == \"myo\":\n",
    "            return \"myocardial disease incl. angina pectoris\"\n",
    "        elif x == \"3005\":\n",
    "            return \"Fracture resulting from simple fall\"\n",
    "        else:\n",
    "            return x\n",
    "\n",
    "\n",
    "def fix_df(df, TRAIT):\n",
    "    if TRAIT == \"BT\":\n",
    "        df[\"OR\"] = np.exp(df.BETA)\n",
    "        df[\"OR_upper\"] = np.exp(df.BETA + df.SE)\n",
    "        df[\"OR_lower\"] = np.exp(df.BETA - df.SE)\n",
    "        df[\"OR_se\"] = df[\"OR_upper\"] - df[\"OR_lower\"]\n",
    "        df[\"OR_se_l\"] = df[\"OR\"] - df[\"OR_lower\"]\n",
    "        df[\"OR_se_u\"] = df[\"OR_upper\"] - df[\"OR\"]\n",
    "    elif TRAIT == \"QT\":\n",
    "        df.loc[:, \"OR\"] = df.BETA\n",
    "        df.loc[:, \"OR_se\"] = df.SE\n",
    "\n",
    "    df.loc[:, \"Phenotype\"] = df.PHENO.apply(lambda x: pheno_search(x))\n",
    "    df.loc[:, \"pval\"] = np.power(10, -df[\"LOG10P\"])\n",
    "    df.loc[:, \"pval_e\"] = df.pval.apply(lambda x: f\"{x:.2f}\")\n",
    "    df.loc[:, \"pval_stars\"] = df[\"pval\"].apply(lambda x: pval(x))\n",
    "    df.loc[:, \"N_pos\"] = (2 * df[\"N\"] * df[\"A1FREQ\"]).astype(int)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c308e029",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [\n",
    "    f\"file:/mnt/project/data/results/{GENE}/{file}\"\n",
    "    for file in os.listdir(f\"/mnt/project/data/results/{GENE}\")\n",
    "]\n",
    "\n",
    "ukb_coding = pd.read_csv(\n",
    "    \"/opt/notebooks/gogoGPCR/data/misc/Data_Dictionary_Showcase.csv\",\n",
    "    error_bad_lines=False,\n",
    "    warn_bad_lines=False,\n",
    "    quotechar='\"',\n",
    ")\n",
    "\n",
    "custom_coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5839ca6f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d467c99d",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "recode = {\n",
    "    \"systolic\": \"Systolic blood pressure\",\n",
    "    \"diastolic\": \"Diastolic blood pressure\",\n",
    "    \"diabetes\": \"Type 2 diabetes\",\n",
    "    \"obesity\": \"Obesity\",\n",
    "    \"blood_pressure_med\": \"Blood pressure medication\",\n",
    "    \"cholesterol_med\": \"Cholesterol medication\",\n",
    "}\n",
    "\n",
    "df = pd.read_csv(files[0], delimiter=\" \", header=\"infer\", comment=\"#\").assign(\n",
    "    SOURCE=os.path.basename(files[0])\n",
    ")\n",
    "df = pd.concat(\n",
    "    [df]\n",
    "    + [\n",
    "        pd.read_csv(fp, delimiter=\" \", comment=\"#\").assign(SOURCE=os.path.basename(fp))\n",
    "        for fp in files[1:]\n",
    "    ],\n",
    "    axis=0,\n",
    ")\n",
    "\n",
    "df.loc[:, \"GENE\"] = df.ID.apply(lambda x: x.split(\".\")[0])\n",
    "df.loc[:, \"MASK\"] = df.ALLELE1.apply(lambda x: x.split(\".\", maxsplit=2)[0])\n",
    "df.loc[:, \"AAF\"] = df.ALLELE1.apply(lambda x: x.split(\".\", maxsplit=1)[-1])\n",
    "df.loc[:, \"FILE\"] = df.SOURCE.apply(\n",
    "    lambda x: x.split(\"/\")[-1].split(\"\\.\")[0].split(\"_\", maxsplit=4)[2:]\n",
    ")\n",
    "df.loc[:, \"TRAIT\"] = df.FILE.apply(lambda x: x[0])\n",
    "df.loc[:, \"PHENO\"] = df.FILE.apply(lambda x: x[-1].split(\".\")[0])\n",
    "df = df.drop([\"ID\", \"ALLELE0\", \"ALLELE1\", \"EXTRA\", \"SOURCE\", \"FILE\", \"TEST\"], axis=1)\n",
    "df.loc[:, \"PHENO\"] = df.loc[:, \"PHENO\"].apply(lambda x: recode.get(x, x))\n",
    "df = df.loc[df[\"AAF\"] != \"singleton\", :]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ad90e12",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Set trait specific\n",
    "TRAIT = \"QT\"\n",
    "\n",
    "titles = {\n",
    "    \"BT\": \"Binary Traits\",\n",
    "    \"QT\": \"Quantitative Traits\",\n",
    "}\n",
    "title = titles[TRAIT]\n",
    "\n",
    "xlabs = {\"BT\": \"OR\", \"QT\": \"Î²\"}\n",
    "xlab = xlabs[TRAIT]\n",
    "\n",
    "masks = df.MASK.unique()\n",
    "\n",
    "# masks = [unique_masks[0], unique_masks[3], unique_masks[1]]\n",
    "\n",
    "interesting_traits = {\n",
    "    \"QT\": [\n",
    "        \"Diastolic blood pressure\",\n",
    "        \"Systolic blood pressure\",\n",
    "        \"Body mass index (BMI)\",\n",
    "        \"Body fat percentage\",\n",
    "    ],\n",
    "    \"BT\": [\n",
    "        \"Type 2 diabetes\",\n",
    "        \"Obesity\",\n",
    "    ],\n",
    "}\n",
    "interesting = interesting_traits[TRAIT]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67977a8e",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c89f6d6",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Plotting options\n",
    "\n",
    "lw = 1.3\n",
    "ms = 10\n",
    "height = 2 * len(interesting)\n",
    "\n",
    "legend_elements = [\n",
    "    Line2D([0], [0], color=\"tab:grey\", marker=\"o\", ms=ms, lw=0, label=\"WT-like\"),\n",
    "    Line2D(\n",
    "        [0], [0], color=\"tab:orange\", marker=\"o\", ms=ms, lw=0, label=\"cAMP LOF only\"\n",
    "    ),\n",
    "    Line2D([0], [0], color=\"tab:purple\", marker=\"o\", ms=ms, lw=0, label=\"Both LOF\"),\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e609aeec",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.loc[df.TRAIT.eq(TRAIT), :]\n",
    "df2 = fix_df(df2, TRAIT=TRAIT)\n",
    "df2 = df2.loc[df2.Phenotype.isin(interesting)]\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95dcaf42",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "plt_df = (\n",
    "    df2.loc[(df2.GENE == GENE)]\n",
    "    .sort_values(by=[\"Phenotype\", \"AAF\"], ascending=[True, False])\n",
    "    .groupby([\"Phenotype\", \"MASK\"])\n",
    "    .first()\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "plt_df = plt_df.loc[\n",
    "    ~plt_df.Phenotype.isin(\n",
    "        [\n",
    "            \"sev_obesity\",\n",
    "            \"Date E66 first reported (obesity)\",\n",
    "            \"Date I10 first reported (essential (primary) hypertension)\",\n",
    "            \"myocardial disease incl. angina pectoris\",\n",
    "            \"Pulse rate, automated reading\",\n",
    "            \"Heart attack diagnosed by doctor\",\n",
    "            \"Angina diagnosed by doctor\",\n",
    "            \"Fracture resulting from simple fall\",\n",
    "            \"Stroke diagnosed by doctor\",\n",
    "        ]\n",
    "    )\n",
    "]\n",
    "\n",
    "group_by_mean = pd.DataFrame(\n",
    "    {\"mean\": plt_df.groupby([\"Phenotype\"]).agg(\"mean\")[\"OR\"]}\n",
    ").reset_index()\n",
    "\n",
    "group_by_mean = group_by_mean.sort_values(by=\"mean\", ascending=False).reset_index()\n",
    "\n",
    "sorter = list(group_by_mean[\"Phenotype\"])\n",
    "\n",
    "plt_df.loc[:, \"Phenotype\"] = plt_df.loc[:, \"Phenotype\"].astype(\"category\")\n",
    "plt_df.loc[:, \"Phenotype\"].cat.set_categories(sorter, inplace=True)\n",
    "\n",
    "plt_df = plt_df.sort_values(\n",
    "    by=[\"Phenotype\", \"MASK\"], ascending=[True, False]\n",
    ").reset_index(drop=True)\n",
    "\n",
    "phenotypes = plt_df.Phenotype.unique()\n",
    "\n",
    "fig, axes = plt.subplots(nrows=len(phenotypes), sharex=True, figsize=(4, height))\n",
    "\n",
    "for ax in range(0, len(phenotypes)):\n",
    "    temp = plt_df.loc[plt_df.Phenotype.eq(phenotypes[ax]), :]\n",
    "    temp = temp.loc[plt_df.MASK.isin(masks)]\n",
    "\n",
    "    temps = [temp[temp.MASK == mask] for mask in masks]\n",
    "\n",
    "    if TRAIT == \"BT\":\n",
    "        xerrs = [[temp[\"OR_se_l\"].values, temp[\"OR_se_u\"].values] for temp in temps]\n",
    "\n",
    "    elif TRAIT == \"QT\":\n",
    "        xerrs = [temp[\"OR_se\"] for temp in temps]\n",
    "\n",
    "    axes[ax].errorbar(\n",
    "        temps[0][\"OR\"],\n",
    "        temps[0].index,\n",
    "        alpha=0.99,\n",
    "        xerr=xerrs[0],\n",
    "        fmt=\"o\",\n",
    "        c=\"tab:grey\",\n",
    "        ecolor=\"black\",\n",
    "        ms=ms,\n",
    "        mew=0.0,\n",
    "        mec=\"black\",\n",
    "        elinewidth=lw,\n",
    "    )\n",
    "    axes[ax].errorbar(\n",
    "        temps[1][\"OR\"],\n",
    "        temps[1].index,\n",
    "        alpha=0.99,\n",
    "        xerr=xerrs[1],\n",
    "        fmt=\"o\",\n",
    "        c=\"tab:purple\",\n",
    "        ecolor=\"black\",\n",
    "        ms=ms,\n",
    "        mew=0.0,\n",
    "        mec=\"black\",\n",
    "        elinewidth=lw,\n",
    "    )\n",
    "    # axes[ax].errorbar(\n",
    "    #    temps[2][\"OR\"],\n",
    "    #    temps[2].index,\n",
    "    #    alpha=0.99,\n",
    "    #    xerr=xerrs[2],\n",
    "    #    fmt=\"o\",\n",
    "    #    c=\"tab:orange\",\n",
    "    #    ecolor=\"black\",\n",
    "    #    ms=ms,\n",
    "    #    mew=0.0,\n",
    "    #    mec=\"black\",\n",
    "    #    elinewidth=lw,\n",
    "    # )\n",
    "\n",
    "    ax0 = axes[ax].twinx()\n",
    "    ax0.set_ylim([0.25, 3.25])\n",
    "    ax0.set_yticks(\n",
    "        [\n",
    "            1,\n",
    "            2,\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    y2labels = (\n",
    "        temp.N_pos.astype(str)\n",
    "        #    + temp.or_e.astype(str).str.ljust(5).values\n",
    "        #    + [\"(\"]\n",
    "        #    + temp.lower_e.values\n",
    "        #    + [\",\"]\n",
    "        #    + temp.upper_e.values\n",
    "        + [\"   p = \"]\n",
    "        + temp.pval_e.values\n",
    "    )\n",
    "\n",
    "    ax0.set_yticklabels(y2labels, fontsize=9, fontdict={\"family\": \"monospace\"})\n",
    "    ax0.tick_params(right=False)\n",
    "    ax0.spines[\"top\"].set_alpha(0)\n",
    "    ax0.spines[\"left\"].set_alpha(0)\n",
    "    ax0.spines[\"right\"].set_alpha(0)\n",
    "    ax0.spines[\"bottom\"].set_alpha(0)\n",
    "    ax0.grid(False)\n",
    "    # axes[ax].invert_xaxis()\n",
    "    # only show every 3rd yticklabel\n",
    "    labels = [l if i % 3 == 0 else \"\" for i, l in enumerate(temp.Phenotype)]\n",
    "    axes[ax].set(yticks=temp.index, yticklabels=labels[::-1])\n",
    "\n",
    "    # axes[ax].axvline(x=0, linestyle=\"--\", color=\"#4f4f4f\")\n",
    "    axes[ax].tick_params(left=False)\n",
    "\n",
    "    if TRAIT == \"BT\":\n",
    "        axes[ax].set_xlim([-0.2, 5.2])\n",
    "        axes[ax].axvline(x=1, linestyle=\":\", color=\"#4f4f4f\")\n",
    "    elif TRAIT == \"QT\":\n",
    "        axes[ax].set_xlim([-0.22, 0.82])\n",
    "        axes[ax].axvline(x=0, linestyle=\":\", color=\"#4f4f4f\")\n",
    "\n",
    "    axes[ax].spines[\"top\"].set_alpha(0)\n",
    "    axes[ax].spines[\"left\"].set_alpha(0)\n",
    "    axes[ax].spines[\"right\"].set_alpha(0)\n",
    "    if ax != len(phenotypes) - 1:\n",
    "        axes[ax].spines[\"bottom\"].set_alpha(0)\n",
    "        axes[ax].tick_params(bottom=False)\n",
    "\n",
    "    if ax == len(phenotypes) - 1:\n",
    "        axes[ax].set_xlabel(xlab)\n",
    "        # ax0.legend(handles = legend_elements, loc = \"lower right\")\n",
    "#\n",
    "# axes[len(phenotypes) - 1].set_xticks([0,1,2,3,4,5,6,7,])\n",
    "for ax in axes.flat:\n",
    "    ax.margins(0.3)\n",
    "\n",
    "plt.subplots_adjust(right=1)\n",
    "# fig.suptitle(title)\n",
    "\n",
    "# plt.savefig(f\"/opt/notebooks/gogoGPCR/tmp/{GENE}_{TRAIT}_new.svg\", dpi = 600, bbox_inches=\"tight\", format = \"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc6e6a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt_df[\"N_pos\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e0734f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"N_alleles\"] = df[\"N\"] * df[\"A1FREQ\"] * 2\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a2a9dbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_search"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "encoding": "# -*- coding: utf-8 -*-",
   "formats": "ipynb,py:percent"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
