{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "60a1d48f",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "import dxdata\n",
    "import dxpy\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyspark\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "import results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "768c05eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flags\n",
    "GENE = \"DRD2\"\n",
    "TRAIT = \"BT\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ab2ffbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sc = pyspark.SparkContext()\n",
    "# spark = pyspark.sql.SparkSession(sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ffdbad15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Results files\n",
    "files = [\n",
    "    f\"file:/mnt/project/Data/results/{file}\"\n",
    "    for file in os.listdir(f\"/mnt/project/Data/results\") if file.endswith(\".regenie\")\n",
    "]\n",
    "\n",
    "# Field codings\n",
    "ukb_coding = pd.read_csv(\n",
    "    \"/opt/notebooks/gogoGPCR/data/misc/Data_Dictionary_Showcase.csv\",\n",
    "    error_bad_lines=False,\n",
    "    warn_bad_lines=False,\n",
    "    quotechar='\"',\n",
    "    usecols=[\"FieldID\", \"Field\"],\n",
    ")\n",
    "\n",
    "custom_coding = pd.read_csv(\n",
    "    \"/opt/notebooks/gogoGPCR/data/misc/Data_Dictionary_Custom.csv\",\n",
    "    error_bad_lines=False,\n",
    "    warn_bad_lines=False,\n",
    "    quotechar='\"',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5e8b99ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load raw DF\n",
    "df_raw = pd.read_csv(files[0], delimiter=\" \", header=\"infer\", comment=\"#\").assign(\n",
    "    SOURCE=os.path.basename(files[0])\n",
    ")\n",
    "df_raw = pd.concat(\n",
    "    [df_raw]\n",
    "    + [\n",
    "        pd.read_csv(fp, delimiter=\" \", comment=\"#\").assign(SOURCE=os.path.basename(fp))\n",
    "        for fp in files[1:]\n",
    "    ],\n",
    "    axis=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "be29360a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>GENPOS</th>\n",
       "      <th>A1FREQ</th>\n",
       "      <th>N</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>CHISQ</th>\n",
       "      <th>LOG10P</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MASK</th>\n",
       "      <th>AAF</th>\n",
       "      <th>TRAIT</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.001470</td>\n",
       "      <td>7.581560</td>\n",
       "      <td>0.017448</td>\n",
       "      <td>0.048220</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>WT</td>\n",
       "      <td>singleton</td>\n",
       "      <td>BT</td>\n",
       "      <td>130840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.004486</td>\n",
       "      <td>180459</td>\n",
       "      <td>-0.602494</td>\n",
       "      <td>0.652356</td>\n",
       "      <td>0.852974</td>\n",
       "      <td>0.448901</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>WT</td>\n",
       "      <td>0.01</td>\n",
       "      <td>BT</td>\n",
       "      <td>130840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.000460</td>\n",
       "      <td>17.269600</td>\n",
       "      <td>0.003356</td>\n",
       "      <td>0.020541</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>LoF</td>\n",
       "      <td>singleton</td>\n",
       "      <td>BT</td>\n",
       "      <td>130840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.024200</td>\n",
       "      <td>1.789000</td>\n",
       "      <td>0.327756</td>\n",
       "      <td>0.246430</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>LoF</td>\n",
       "      <td>0.01</td>\n",
       "      <td>BT</td>\n",
       "      <td>130840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.072390</td>\n",
       "      <td>1.937940</td>\n",
       "      <td>0.306214</td>\n",
       "      <td>0.236563</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>WT</td>\n",
       "      <td>singleton</td>\n",
       "      <td>BT</td>\n",
       "      <td>130846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM     GENPOS    A1FREQ       N      BETA         SE     CHISQ  \\\n",
       "0     11  113410735  0.000064  180459 -1.001470   7.581560  0.017448   \n",
       "1     11  113410735  0.004486  180459 -0.602494   0.652356  0.852974   \n",
       "2     11  113410735  0.000019  180459 -1.000460  17.269600  0.003356   \n",
       "3     11  113410735  0.000413  180459 -1.024200   1.789000  0.327756   \n",
       "0     11  113410735  0.000064  180459 -1.072390   1.937940  0.306214   \n",
       "\n",
       "     LOG10P  GENE MASK        AAF TRAIT   PHENO  \n",
       "0  0.048220  DRD2   WT  singleton    BT  130840  \n",
       "1  0.448901  DRD2   WT       0.01    BT  130840  \n",
       "2  0.020541  DRD2  LoF  singleton    BT  130840  \n",
       "3  0.246430  DRD2  LoF       0.01    BT  130840  \n",
       "0  0.236563  DRD2   WT  singleton    BT  130846  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Fix common fields\n",
    "df = df_raw\n",
    "df.loc[:, \"GENE\"] = df.ID.apply(lambda x: x.split(\".\")[0])\n",
    "df.loc[:, \"MASK\"] = df.ALLELE1.apply(lambda x: x.split(\".\", maxsplit=2)[0])\n",
    "df.loc[:, \"AAF\"] = df.ALLELE1.apply(lambda x: x.split(\".\", maxsplit=1)[-1])\n",
    "# df.loc[:,\"MASK2\"] = df.MASK + \".\" + df.AAF\n",
    "df.loc[:, \"FILE\"] = df.SOURCE.apply(\n",
    "    lambda x: x.split(\"/\")[-1].split(\"\\.\")[0].split(\"_\", maxsplit=4)[2:]\n",
    ")\n",
    "df.loc[:, \"TRAIT\"] = TRAIT # df.FILE.apply(lambda x: x[0])\n",
    "df.loc[:, \"PHENO\"] = df.FILE.apply(lambda x: x[-1].split(\".\")[0])\n",
    "df = df.drop([\"ID\", \"ALLELE0\", \"ALLELE1\", \"EXTRA\", \"SOURCE\", \"FILE\", \"TEST\"], axis=1)\n",
    "\n",
    "# Sanity check\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2b456ddd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>GENPOS</th>\n",
       "      <th>A1FREQ</th>\n",
       "      <th>N</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>CHISQ</th>\n",
       "      <th>LOG10P</th>\n",
       "      <th>GENE</th>\n",
       "      <th>MASK</th>\n",
       "      <th>...</th>\n",
       "      <th>OR_up</th>\n",
       "      <th>OR_low</th>\n",
       "      <th>OR_up_lim</th>\n",
       "      <th>OR_low_lim</th>\n",
       "      <th>BETA_up_lim</th>\n",
       "      <th>BETA_low_lim</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>pval</th>\n",
       "      <th>pval_stars</th>\n",
       "      <th>N_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.001470</td>\n",
       "      <td>7.581560</td>\n",
       "      <td>0.017448</td>\n",
       "      <td>0.048220</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>7.206042e+02</td>\n",
       "      <td>1.872567e-04</td>\n",
       "      <td>7.202368e+02</td>\n",
       "      <td>0.367152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Date F02 first reported (dementia in other dis...</td>\n",
       "      <td>0.894911</td>\n",
       "      <td></td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.004486</td>\n",
       "      <td>180459</td>\n",
       "      <td>-0.602494</td>\n",
       "      <td>0.652356</td>\n",
       "      <td>0.852974</td>\n",
       "      <td>0.448901</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>1.051126e+00</td>\n",
       "      <td>2.851186e-01</td>\n",
       "      <td>5.036814e-01</td>\n",
       "      <td>0.262326</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Date F02 first reported (dementia in other dis...</td>\n",
       "      <td>0.355712</td>\n",
       "      <td></td>\n",
       "      <td>1618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.000460</td>\n",
       "      <td>17.269600</td>\n",
       "      <td>0.003356</td>\n",
       "      <td>0.020541</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>LoF</td>\n",
       "      <td>...</td>\n",
       "      <td>1.163048e+07</td>\n",
       "      <td>1.162556e-08</td>\n",
       "      <td>1.163048e+07</td>\n",
       "      <td>0.367710</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Date F02 first reported (dementia in other dis...</td>\n",
       "      <td>0.953803</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.024200</td>\n",
       "      <td>1.789000</td>\n",
       "      <td>0.327756</td>\n",
       "      <td>0.246430</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>LoF</td>\n",
       "      <td>...</td>\n",
       "      <td>2.148565e+00</td>\n",
       "      <td>6.001264e-02</td>\n",
       "      <td>1.789481e+00</td>\n",
       "      <td>0.299071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Date F02 first reported (dementia in other dis...</td>\n",
       "      <td>0.566983</td>\n",
       "      <td></td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.072390</td>\n",
       "      <td>1.937940</td>\n",
       "      <td>0.306214</td>\n",
       "      <td>0.236563</td>\n",
       "      <td>DRD2</td>\n",
       "      <td>WT</td>\n",
       "      <td>...</td>\n",
       "      <td>2.376313e+00</td>\n",
       "      <td>4.927542e-02</td>\n",
       "      <td>2.034123e+00</td>\n",
       "      <td>0.292914</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Date F05 first reported (delirium, not induced...</td>\n",
       "      <td>0.580012</td>\n",
       "      <td></td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM     GENPOS    A1FREQ       N      BETA         SE     CHISQ  \\\n",
       "0     11  113410735  0.000064  180459 -1.001470   7.581560  0.017448   \n",
       "1     11  113410735  0.004486  180459 -0.602494   0.652356  0.852974   \n",
       "2     11  113410735  0.000019  180459 -1.000460  17.269600  0.003356   \n",
       "3     11  113410735  0.000413  180459 -1.024200   1.789000  0.327756   \n",
       "0     11  113410735  0.000064  180459 -1.072390   1.937940  0.306214   \n",
       "\n",
       "     LOG10P  GENE MASK  ...         OR_up        OR_low     OR_up_lim  \\\n",
       "0  0.048220  DRD2   WT  ...  7.206042e+02  1.872567e-04  7.202368e+02   \n",
       "1  0.448901  DRD2   WT  ...  1.051126e+00  2.851186e-01  5.036814e-01   \n",
       "2  0.020541  DRD2  LoF  ...  1.163048e+07  1.162556e-08  1.163048e+07   \n",
       "3  0.246430  DRD2  LoF  ...  2.148565e+00  6.001264e-02  1.789481e+00   \n",
       "0  0.236563  DRD2   WT  ...  2.376313e+00  4.927542e-02  2.034123e+00   \n",
       "\n",
       "   OR_low_lim  BETA_up_lim  BETA_low_lim  \\\n",
       "0    0.367152          NaN           NaN   \n",
       "1    0.262326          NaN           NaN   \n",
       "2    0.367710          NaN           NaN   \n",
       "3    0.299071          NaN           NaN   \n",
       "0    0.292914          NaN           NaN   \n",
       "\n",
       "                                           Phenotype      pval  pval_stars  \\\n",
       "0  Date F02 first reported (dementia in other dis...  0.894911               \n",
       "1  Date F02 first reported (dementia in other dis...  0.355712               \n",
       "2  Date F02 first reported (dementia in other dis...  0.953803               \n",
       "3  Date F02 first reported (dementia in other dis...  0.566983               \n",
       "0  Date F05 first reported (delirium, not induced...  0.580012               \n",
       "\n",
       "   N_pos  \n",
       "0     23  \n",
       "1   1618  \n",
       "2      7  \n",
       "3    148  \n",
       "0     23  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filters\n",
    "bt = df.TRAIT == \"BT\"\n",
    "qt = df.TRAIT == \"QT\"\n",
    "\n",
    "# Fix Binary Traits\n",
    "df.loc[bt, \"OR\"] = np.exp(df.loc[bt, \"BETA\"])\n",
    "df.loc[bt, \"OR_up\"] = np.exp(df.loc[bt, \"BETA\"] + df.loc[bt, \"SE\"])\n",
    "df.loc[bt, \"OR_low\"] = np.exp(df.loc[bt, \"BETA\"] - df.loc[bt, \"SE\"])\n",
    "df.loc[bt, \"OR_up_lim\"] = df.loc[bt, \"OR_up\"] - df.loc[bt, \"OR\"]\n",
    "df.loc[bt, \"OR_low_lim\"] = df.loc[bt, \"OR\"] - df.loc[bt, \"OR_low\"]\n",
    "\n",
    "# Fix Quantitative Traits\n",
    "df.loc[qt, \"BETA_up_lim\"] = df.loc[qt, \"BETA\"] + df.loc[qt, \"SE\"]\n",
    "df.loc[qt, \"BETA_low_lim\"] = df.loc[qt, \"BETA\"] - df.loc[qt, \"SE\"]\n",
    "\n",
    "# Final fixes\n",
    "df.loc[:, \"Phenotype\"] = df.PHENO.apply(\n",
    "    lambda x: results.pheno_search(x, ukb_coding, custom_coding)\n",
    "    .replace('\"', \"\")\n",
    "    .strip()\n",
    ")\n",
    "df.loc[:, \"pval\"] = np.power(10, -df[\"LOG10P\"])\n",
    "df.loc[:, \"pval_stars\"] = df[\"pval\"].apply(lambda x: results.pval_stars(x))\n",
    "df.loc[:, \"N_pos\"] = (2 * df[\"N\"] * df[\"A1FREQ\"]).astype(int)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4b754060",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "phenos_to_remove = [\n",
    "    \"Severe obesity\",\n",
    "    \"Date E66 first reported (obesity)\",\n",
    "    \"Date I10 first reported (essential (primary) hypertension)\",\n",
    "    \"Myocardial disease incl. angina pectoris\",\n",
    "    \"Pulse rate, automated reading\",\n",
    "    \"Heart attack diagnosed by doctor\",\n",
    "    \"Angina diagnosed by doctor\",\n",
    "    \"Fracture resulting from simple fall\",\n",
    "    \"Stroke diagnosed by doctor\",\n",
    "    \"Cholesterol medication\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "552c8c7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>MASK</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>GENPOS</th>\n",
       "      <th>A1FREQ</th>\n",
       "      <th>N</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>CHISQ</th>\n",
       "      <th>LOG10P</th>\n",
       "      <th>...</th>\n",
       "      <th>OR_up</th>\n",
       "      <th>OR_low</th>\n",
       "      <th>OR_up_lim</th>\n",
       "      <th>OR_low_lim</th>\n",
       "      <th>BETA_up_lim</th>\n",
       "      <th>BETA_low_lim</th>\n",
       "      <th>pval</th>\n",
       "      <th>pval_stars</th>\n",
       "      <th>N_pos</th>\n",
       "      <th>Phenotypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bipolar I Disorder</td>\n",
       "      <td>WT</td>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>54564</td>\n",
       "      <td>-1.02760</td>\n",
       "      <td>4.14654</td>\n",
       "      <td>0.061415</td>\n",
       "      <td>0.094596</td>\n",
       "      <td>...</td>\n",
       "      <td>22.622387</td>\n",
       "      <td>0.005661</td>\n",
       "      <td>22.264522</td>\n",
       "      <td>0.352204</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.804274</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>Bipolar I Disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bipolar I Disorder</td>\n",
       "      <td>LoF</td>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>54564</td>\n",
       "      <td>-1.01121</td>\n",
       "      <td>1.98287</td>\n",
       "      <td>0.260073</td>\n",
       "      <td>0.214621</td>\n",
       "      <td>...</td>\n",
       "      <td>2.642327</td>\n",
       "      <td>0.050083</td>\n",
       "      <td>2.278549</td>\n",
       "      <td>0.313696</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.610069</td>\n",
       "      <td></td>\n",
       "      <td>59</td>\n",
       "      <td>Bipolar I Disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bipolar II Disorder</td>\n",
       "      <td>WT</td>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>54564</td>\n",
       "      <td>-1.03839</td>\n",
       "      <td>3.76719</td>\n",
       "      <td>0.075978</td>\n",
       "      <td>0.106336</td>\n",
       "      <td>...</td>\n",
       "      <td>15.314499</td>\n",
       "      <td>0.008184</td>\n",
       "      <td>14.960474</td>\n",
       "      <td>0.345840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.782824</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>Bipolar II Disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bipolar II Disorder</td>\n",
       "      <td>LoF</td>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>54564</td>\n",
       "      <td>-1.01326</td>\n",
       "      <td>1.68678</td>\n",
       "      <td>0.360846</td>\n",
       "      <td>0.261190</td>\n",
       "      <td>...</td>\n",
       "      <td>1.961128</td>\n",
       "      <td>0.067203</td>\n",
       "      <td>1.598095</td>\n",
       "      <td>0.295831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.548037</td>\n",
       "      <td></td>\n",
       "      <td>59</td>\n",
       "      <td>Bipolar II Disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Date F02 first reported (dementia in other dis...</td>\n",
       "      <td>WT</td>\n",
       "      <td>11</td>\n",
       "      <td>113410735</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>180459</td>\n",
       "      <td>-1.00147</td>\n",
       "      <td>7.58156</td>\n",
       "      <td>0.017448</td>\n",
       "      <td>0.048220</td>\n",
       "      <td>...</td>\n",
       "      <td>720.604181</td>\n",
       "      <td>0.000187</td>\n",
       "      <td>720.236842</td>\n",
       "      <td>0.367152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.894911</td>\n",
       "      <td></td>\n",
       "      <td>23</td>\n",
       "      <td>Date F02 first reported (dementia in other dis...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Phenotype MASK  CHROM     GENPOS  \\\n",
       "0                                 Bipolar I Disorder   WT     11  113410735   \n",
       "1                                 Bipolar I Disorder  LoF     11  113410735   \n",
       "2                                Bipolar II Disorder   WT     11  113410735   \n",
       "3                                Bipolar II Disorder  LoF     11  113410735   \n",
       "4  Date F02 first reported (dementia in other dis...   WT     11  113410735   \n",
       "\n",
       "     A1FREQ       N     BETA       SE     CHISQ    LOG10P  ...       OR_up  \\\n",
       "0  0.000064   54564 -1.02760  4.14654  0.061415  0.094596  ...   22.622387   \n",
       "1  0.000541   54564 -1.01121  1.98287  0.260073  0.214621  ...    2.642327   \n",
       "2  0.000064   54564 -1.03839  3.76719  0.075978  0.106336  ...   15.314499   \n",
       "3  0.000541   54564 -1.01326  1.68678  0.360846  0.261190  ...    1.961128   \n",
       "4  0.000064  180459 -1.00147  7.58156  0.017448  0.048220  ...  720.604181   \n",
       "\n",
       "     OR_low   OR_up_lim OR_low_lim  BETA_up_lim  BETA_low_lim      pval  \\\n",
       "0  0.005661   22.264522   0.352204          NaN           NaN  0.804274   \n",
       "1  0.050083    2.278549   0.313696          NaN           NaN  0.610069   \n",
       "2  0.008184   14.960474   0.345840          NaN           NaN  0.782824   \n",
       "3  0.067203    1.598095   0.295831          NaN           NaN  0.548037   \n",
       "4  0.000187  720.236842   0.367152          NaN           NaN  0.894911   \n",
       "\n",
       "   pval_stars  N_pos                                         Phenotypes  \n",
       "0                  7                                 Bipolar I Disorder  \n",
       "1                 59                                 Bipolar I Disorder  \n",
       "2                  7                                Bipolar II Disorder  \n",
       "3                 59                                Bipolar II Disorder  \n",
       "4                 23  Date F02 first reported (dementia in other dis...  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plt_df = (\n",
    "    df.loc[(df.GENE == GENE)]\n",
    "    .sort_values(by=[\"Phenotype\", \"AAF\"], ascending=[True, False])\n",
    "    .groupby([\"Phenotype\", \"MASK\"])\n",
    "    .first()\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "\n",
    "plt_df = plt_df.loc[~plt_df.Phenotype.astype(str).isin(phenos_to_remove), :]\n",
    "\n",
    "effect = {\"BT\": \"OR\", \"QT\": \"BETA\"}[TRAIT]\n",
    "\n",
    "group_by_mean = (\n",
    "    pd.DataFrame({\"mean\": plt_df.groupby([\"Phenotype\"]).agg(\"mean\")[effect]})\n",
    "    .sort_values(by=\"mean\", ascending=False)\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "sorter = group_by_mean.Phenotype.tolist()\n",
    "\n",
    "plt_df.loc[:, \"Phenotype\"] = plt_df.loc[:, \"Phenotype\"].astype(\"category\")\n",
    "plt_df.loc[:, \"Phenotypes\"] = plt_df.loc[:, \"Phenotype\"].cat.set_categories(sorter)\n",
    "\n",
    "plt_df = plt_df.sort_values(\n",
    "    by=[\"Phenotype\", \"MASK\"], ascending=[True, False]\n",
    ").reset_index(drop=True)\n",
    "\n",
    "phenotypes = plt_df.Phenotype.unique()\n",
    "\n",
    "print(len(plt_df))\n",
    "plt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a480b928",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_BT(\n",
    "    df, width=4, height=12, phenotypes=None, masks=None, lw=1.3, ms=10, fudge=0.6\n",
    "):\n",
    "\n",
    "    df = df.loc[df.TRAIT == \"BT\", :]\n",
    "\n",
    "    df.loc[:, \"CI\"] = +(\n",
    "        \"(\"\n",
    "        + df.loc[:, \"OR_low\"].round(2).astype(str)\n",
    "        + \",\"\n",
    "        + df.loc[:, \"OR_up\"].round(2).astype(str)\n",
    "        + \")\"\n",
    "    )\n",
    "\n",
    "    just = df.applymap(lambda x: len(str(x)) + 1).max()\n",
    "\n",
    "    df.loc[:, \"label\"] = (\n",
    "        df.loc[:, \"N_pos\"].astype(str).str.ljust(just[\"N_pos\"])\n",
    "        + df.loc[:, \"OR\"].round(2).astype(str).str.ljust(5)\n",
    "        + df.loc[:, \"CI\"].str.ljust(just[\"CI\"])\n",
    "        + \"p= \"\n",
    "        + df.loc[:, \"pval\"].apply(lambda p: f\"{p:.2e}\").str.ljust(4)\n",
    "    )\n",
    "\n",
    "    if phenotypes is None:\n",
    "        phenotypes = df.Phenotype.unique()\n",
    "\n",
    "    num_pheno = len(phenotypes)\n",
    "\n",
    "    height = 2 * num_pheno\n",
    "\n",
    "    masks = df.MASK.unique()\n",
    "    colors = plt.cm.viridis(np.linspace(0, 1, len(masks)))\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=num_pheno, sharex=True, figsize=(width, height))\n",
    "    legend_elements = [\n",
    "        Line2D([0], [0], color=color, marker=\"o\", ms=ms, lw=0, label=mask)\n",
    "        for color, mask in zip(colors, masks)\n",
    "    ]\n",
    "\n",
    "    for i, ax in enumerate(axes):\n",
    "\n",
    "        temp = df.loc[df.Phenotype.eq(phenotypes[i]), :]\n",
    "        temp = temp.loc[df.MASK.isin(masks)]\n",
    "\n",
    "        for mask, color in zip(\n",
    "            masks,\n",
    "            colors,\n",
    "        ):\n",
    "            temp1 = temp.loc[temp.MASK == mask, :]\n",
    "            xerr = [temp1[\"OR_low_lim\"].values, temp1[\"OR_up_lim\"].values]\n",
    "\n",
    "            ax.errorbar(\n",
    "                temp1[\"OR\"],\n",
    "                temp1.index,\n",
    "                alpha=0.99,\n",
    "                xerr=xerr,\n",
    "                fmt=\"o\",\n",
    "                c=color,\n",
    "                ecolor=\"black\",\n",
    "                ms=ms,\n",
    "                mew=0.0,\n",
    "                mec=\"black\",\n",
    "                elinewidth=lw,\n",
    "            )\n",
    "\n",
    "        ax0 = ax.twinx()\n",
    "        ax0.yaxis.tick_left()\n",
    "        ax.yaxis.tick_right()\n",
    "        ax.set_ylim(temp.index[0] - fudge, temp.index[1] + fudge)\n",
    "        ax.set_yticks([temp.index[0], temp.index[1]])\n",
    "        ax.set_xlim(0 - fudge, df.OR_up.max() + fudge)\n",
    "        ax.set_xticks(np.arange(0, int(df.OR_up.max()) + 2, 1))\n",
    "\n",
    "        ax.set_yticklabels(temp.label, fontsize=9, fontdict={\"family\": \"monospace\"})\n",
    "\n",
    "        ax.tick_params(right=False)\n",
    "        ax.spines[\"top\"].set_alpha(0)\n",
    "        ax.spines[\"left\"].set_alpha(0)\n",
    "        ax.spines[\"right\"].set_alpha(0)\n",
    "        ax.spines[\"bottom\"].set_alpha(0)\n",
    "\n",
    "        ax0.tick_params(right=False)\n",
    "        ax0.tick_params(left=False)\n",
    "        ax0.spines[\"top\"].set_alpha(0)\n",
    "        ax0.spines[\"left\"].set_alpha(0)\n",
    "        ax0.spines[\"right\"].set_alpha(0)\n",
    "\n",
    "        ax0.grid(False)\n",
    "\n",
    "        # only show every 3rd yticklabel\n",
    "        labels = [\n",
    "            l if i % len(masks) == 0 else \"\" for i, l in enumerate(temp.Phenotype)\n",
    "        ]\n",
    "        ax0.set(yticks=temp.index, yticklabels=labels[::-1])\n",
    "\n",
    "        ax.axvline(x=1, linestyle=\"--\", color=\"#4f4f4f\")\n",
    "\n",
    "        ax0.spines[\"top\"].set_alpha(0)\n",
    "        ax0.spines[\"left\"].set_alpha(0)\n",
    "        ax0.spines[\"right\"].set_alpha(0)\n",
    "\n",
    "        if i != len(phenotypes) - 1:\n",
    "            ax0.spines[\"bottom\"].set_alpha(0)\n",
    "            ax.tick_params(bottom=False)\n",
    "\n",
    "        if i == len(phenotypes) - 1:\n",
    "            ax.set_xlabel(\"OR\")\n",
    "            ax.tick_params(bottom=True)\n",
    "            ax.legend(handles=legend_elements, loc=\"lower right\")\n",
    "\n",
    "    return temp\n",
    "\n",
    "\n",
    "temp = plot_BT(plt_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25038dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9075070a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>MASK</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>GENPOS</th>\n",
       "      <th>A1FREQ</th>\n",
       "      <th>N</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>CHISQ</th>\n",
       "      <th>LOG10P</th>\n",
       "      <th>...</th>\n",
       "      <th>OR</th>\n",
       "      <th>OR_up</th>\n",
       "      <th>OR_low</th>\n",
       "      <th>OR_up_lim</th>\n",
       "      <th>OR_low_lim</th>\n",
       "      <th>BETA_up_lim</th>\n",
       "      <th>BETA_low_lim</th>\n",
       "      <th>pval</th>\n",
       "      <th>pval_stars</th>\n",
       "      <th>N_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Diastolic blood pressure</td>\n",
       "      <td>WT</td>\n",
       "      <td>11</td>\n",
       "      <td>113410736</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>146503</td>\n",
       "      <td>0.441296</td>\n",
       "      <td>0.255072</td>\n",
       "      <td>2.993190</td>\n",
       "      <td>1.077710</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.696368</td>\n",
       "      <td>0.186224</td>\n",
       "      <td>0.083616</td>\n",
       "      <td></td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Diastolic blood pressure</td>\n",
       "      <td>LoF</td>\n",
       "      <td>11</td>\n",
       "      <td>113410736</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>146503</td>\n",
       "      <td>0.730394</td>\n",
       "      <td>0.426809</td>\n",
       "      <td>2.928520</td>\n",
       "      <td>1.060350</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.157203</td>\n",
       "      <td>0.303585</td>\n",
       "      <td>0.087026</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fractured/broken bones in last 5 years</td>\n",
       "      <td>WT</td>\n",
       "      <td>11</td>\n",
       "      <td>113410736</td>\n",
       "      <td>0.000047</td>\n",
       "      <td>149631</td>\n",
       "      <td>0.099278</td>\n",
       "      <td>0.934201</td>\n",
       "      <td>0.011293</td>\n",
       "      <td>0.038404</td>\n",
       "      <td>...</td>\n",
       "      <td>1.104373</td>\n",
       "      <td>2.810826</td>\n",
       "      <td>0.433908</td>\n",
       "      <td>1.706454</td>\n",
       "      <td>0.670465</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.915368</td>\n",
       "      <td></td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fractured/broken bones in last 5 years</td>\n",
       "      <td>LoF</td>\n",
       "      <td>11</td>\n",
       "      <td>113410736</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>149631</td>\n",
       "      <td>0.912393</td>\n",
       "      <td>1.387560</td>\n",
       "      <td>0.432373</td>\n",
       "      <td>0.291726</td>\n",
       "      <td>...</td>\n",
       "      <td>2.490275</td>\n",
       "      <td>9.973714</td>\n",
       "      <td>0.621781</td>\n",
       "      <td>7.483439</td>\n",
       "      <td>1.868493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.510827</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Obesity</td>\n",
       "      <td>WT</td>\n",
       "      <td>11</td>\n",
       "      <td>113410736</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>85706</td>\n",
       "      <td>0.440727</td>\n",
       "      <td>0.813380</td>\n",
       "      <td>0.293597</td>\n",
       "      <td>0.230678</td>\n",
       "      <td>...</td>\n",
       "      <td>1.553836</td>\n",
       "      <td>3.504707</td>\n",
       "      <td>0.688904</td>\n",
       "      <td>1.950871</td>\n",
       "      <td>0.864932</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.587925</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Phenotype MASK  CHROM     GENPOS    A1FREQ  \\\n",
       "0                Diastolic blood pressure   WT     11  113410736  0.000048   \n",
       "1                Diastolic blood pressure  LoF     11  113410736  0.000017   \n",
       "2  Fractured/broken bones in last 5 years   WT     11  113410736  0.000047   \n",
       "3  Fractured/broken bones in last 5 years  LoF     11  113410736  0.000017   \n",
       "4                                 Obesity   WT     11  113410736  0.000041   \n",
       "\n",
       "        N      BETA        SE     CHISQ    LOG10P  ...        OR     OR_up  \\\n",
       "0  146503  0.441296  0.255072  2.993190  1.077710  ...       NaN       NaN   \n",
       "1  146503  0.730394  0.426809  2.928520  1.060350  ...       NaN       NaN   \n",
       "2  149631  0.099278  0.934201  0.011293  0.038404  ...  1.104373  2.810826   \n",
       "3  149631  0.912393  1.387560  0.432373  0.291726  ...  2.490275  9.973714   \n",
       "4   85706  0.440727  0.813380  0.293597  0.230678  ...  1.553836  3.504707   \n",
       "\n",
       "     OR_low OR_up_lim  OR_low_lim  BETA_up_lim  BETA_low_lim      pval  \\\n",
       "0       NaN       NaN         NaN     0.696368      0.186224  0.083616   \n",
       "1       NaN       NaN         NaN     1.157203      0.303585  0.087026   \n",
       "2  0.433908  1.706454    0.670465          NaN           NaN  0.915368   \n",
       "3  0.621781  7.483439    1.868493          NaN           NaN  0.510827   \n",
       "4  0.688904  1.950871    0.864932          NaN           NaN  0.587925   \n",
       "\n",
       "   pval_stars  N_pos  \n",
       "0                 14  \n",
       "1                  5  \n",
       "2                 13  \n",
       "3                  5  \n",
       "4                  7  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def make_plt_df(\n",
    "    df: pd.DataFrame,\n",
    "    phenos_to_remove,\n",
    "    trait=TRAIT,\n",
    "    gene=GENE,\n",
    ") -> pd.DataFrame:\n",
    "\n",
    "    plt_df = (\n",
    "        df.loc[(df.GENE == GENE)]\n",
    "        .sort_values(by=[\"Phenotype\", \"AAF\"], ascending=[True, False])\n",
    "        .groupby([\"Phenotype\", \"MASK\"])\n",
    "        .first()\n",
    "        .reset_index()\n",
    "    )\n",
    "\n",
    "    plt_df = plt_df.loc[~plt_df.Phenotype.isin(phenos_to_remove), :]\n",
    "\n",
    "    effect = {\"BT\": \"OR\", \"QT\": \"BETA\"}[TRAIT]\n",
    "\n",
    "    group_by_mean = pd.DataFrame(\n",
    "        {\"mean\": plt_df.groupby([\"Phenotype\"]).agg(\"mean\")[effect]}\n",
    "    ).reset_index()\n",
    "\n",
    "    group_by_mean = group_by_mean.sort_values(by=\"mean\", ascending=False).reset_index()\n",
    "\n",
    "    sorter = list(group_by_mean[\"Phenotype\"])\n",
    "\n",
    "    plt_df.loc[:, \"Phenotype\"] = plt_df.loc[:, \"Phenotype\"].astype(\"category\")\n",
    "    plt_df.loc[:, \"Phenotype\"].cat.set_categories(sorter, inplace=True)\n",
    "\n",
    "    plt_df = plt_df.sort_values(\n",
    "        by=[\"Phenotype\", \"MASK\"], ascending=[True, False]\n",
    "    ).reset_index(drop=True)\n",
    "\n",
    "    phenotypes = plt_df.Phenotype.unique()\n",
    "\n",
    "    return plt_df\n",
    "\n",
    "\n",
    "plt_df = make_plt_df(df, phenos_to_remove=phenos_to_remove)\n",
    "plt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "797867ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ax in range(0, len(phenotypes)):\n",
    "    temp = plt_df.loc[plt_df.Phenotype.eq(phenotypes[ax]), :]\n",
    "    temp = temp.loc[plt_df.MASK.isin(masks)]\n",
    "\n",
    "    temps = [temp[temp.MASK == mask] for mask in masks]\n",
    "\n",
    "    if TRAIT == \"BT\":\n",
    "        xerrs = [[temp[\"OR_se_l\"].values, temp[\"OR_se_u\"].values] for temp in temps]\n",
    "\n",
    "    elif TRAIT == \"QT\":\n",
    "        xerrs = [temp[\"OR_se\"] for temp in temps]\n",
    "\n",
    "    axes[ax].errorbar(\n",
    "        temps[0][\"OR\"],\n",
    "        temps[0].index,\n",
    "        alpha=0.99,\n",
    "        xerr=xerrs[0],\n",
    "        fmt=\"o\",\n",
    "        c=\"tab:grey\",\n",
    "        ecolor=\"black\",\n",
    "        ms=ms,\n",
    "        mew=0.0,\n",
    "        mec=\"black\",\n",
    "        elinewidth=lw,\n",
    "    )\n",
    "    axes[ax].errorbar(\n",
    "        temps[1][\"OR\"],\n",
    "        temps[1].index,\n",
    "        alpha=0.99,\n",
    "        xerr=xerrs[1],\n",
    "        fmt=\"o\",\n",
    "        c=\"tab:purple\",\n",
    "        ecolor=\"black\",\n",
    "        ms=ms,\n",
    "        mew=0.0,\n",
    "        mec=\"black\",\n",
    "        elinewidth=lw,\n",
    "    )\n",
    "    # axes[ax].errorbar(\n",
    "    #    temps[2][\"OR\"],\n",
    "    #    temps[2].index,\n",
    "    #    alpha=0.99,\n",
    "    #    xerr=xerrs[2],\n",
    "    #    fmt=\"o\",\n",
    "    #    c=\"tab:orange\",\n",
    "    #    ecolor=\"black\",\n",
    "    #    ms=ms,\n",
    "    #    mew=0.0,\n",
    "    #    mec=\"black\",\n",
    "    #    elinewidth=lw,\n",
    "    # )\n",
    "\n",
    "    ax0 = axes[ax].twinx()\n",
    "    ax0.set_ylim([0.25, 3.25])\n",
    "    ax0.set_yticks(\n",
    "        [\n",
    "            1,\n",
    "            2,\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    y2labels = (\n",
    "        temp.N_pos.astype(str)\n",
    "        #    + temp.or_e.astype(str).str.ljust(5).values\n",
    "        #    + [\"(\"]\n",
    "        #    + temp.lower_e.values\n",
    "        #    + [\",\"]\n",
    "        #    + temp.upper_e.values\n",
    "        + [\"   p = \"]\n",
    "        + temp.pval_e.values\n",
    "    )\n",
    "\n",
    "    ax0.set_yticklabels(y2labels, fontsize=9, fontdict={\"family\": \"monospace\"})\n",
    "    ax0.tick_params(right=False)\n",
    "    ax0.spines[\"top\"].set_alpha(0)\n",
    "    ax0.spines[\"left\"].set_alpha(0)\n",
    "    ax0.spines[\"right\"].set_alpha(0)\n",
    "    ax0.spines[\"bottom\"].set_alpha(0)\n",
    "    ax0.grid(False)\n",
    "    # axes[ax].invert_xaxis()\n",
    "    # only show every 3rd yticklabel\n",
    "    labels = [l if i % 3 == 0 else \"\" for i, l in enumerate(temp.Phenotype)]\n",
    "    axes[ax].set(yticks=temp.index, yticklabels=labels[::-1])\n",
    "\n",
    "    # axes[ax].axvline(x=0, linestyle=\"--\", color=\"#4f4f4f\")\n",
    "    axes[ax].tick_params(left=False)\n",
    "\n",
    "    if TRAIT == \"BT\":\n",
    "        axes[ax].set_xlim([-0.2, 5.2])\n",
    "        axes[ax].axvline(x=1, linestyle=\":\", color=\"#4f4f4f\")\n",
    "    elif TRAIT == \"QT\":\n",
    "        axes[ax].set_xlim([-0.22, 0.82])\n",
    "        axes[ax].axvline(x=0, linestyle=\":\", color=\"#4f4f4f\")\n",
    "\n",
    "    axes[ax].spines[\"top\"].set_alpha(0)\n",
    "    axes[ax].spines[\"left\"].set_alpha(0)\n",
    "    axes[ax].spines[\"right\"].set_alpha(0)\n",
    "    if ax != len(phenotypes) - 1:\n",
    "        axes[ax].spines[\"bottom\"].set_alpha(0)\n",
    "        axes[ax].tick_params(bottom=False)\n",
    "\n",
    "    if ax == len(phenotypes) - 1:\n",
    "        axes[ax].set_xlabel(xlab)\n",
    "        # ax0.legend(handles = legend_elements, loc = \"lower right\")\n",
    "#\n",
    "# axes[len(phenotypes) - 1].set_xticks([0,1,2,3,4,5,6,7,])\n",
    "for ax in axes.flat:\n",
    "    ax.margins(0.3)\n",
    "\n",
    "plt.subplots_adjust(right=1)\n",
    "# fig.suptitle(title)\n",
    "\n",
    "# plt.savefig(f\"/opt/notebooks/gogoGPCR/tmp/{GENE}_{TRAIT}_new.svg\", dpi = 600, bbox_inches=\"tight\", format = \"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3e05060",
   "metadata": {},
   "outputs": [],
   "source": [
    "###\n",
    "### ALL THE OLD SHIT\n",
    "###\n",
    "# -*- coding: utf-8 -*-\n",
    "# ---\n",
    "# jupyter:\n",
    "#   jupytext:\n",
    "#     formats: ipynb,py:percent\n",
    "#     text_representation:\n",
    "#       extension: .py\n",
    "#       format_name: percent\n",
    "#       format_version: '1.3'\n",
    "#       jupytext_version: 1.13.1\n",
    "#   kernelspec:\n",
    "#     display_name: Python 3\n",
    "#     language: python\n",
    "#     name: python3\n",
    "# ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf92234",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# !pip install -U matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d86107",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import dxdata\n",
    "import dxpy\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyspark\n",
    "import tomli\n",
    "from matplotlib.lines import Line2D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27cd8d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../config.toml\") as f:\n",
    "    conf = tomli.load(f)\n",
    "\n",
    "GENE = conf[\"GENE\"]\n",
    "# REFERENCE_GENOME = conf[\"REFERENCE_GENOME\"]\n",
    "# LOG_FILE = Path(conf[\"IMPORT\"][\"LOG_DIR\"], f\"{GENE}_results.log\").resolve().__str__()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f52c1f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Spark and Hail\n",
    "sc = pyspark.SparkContext()\n",
    "spark = pyspark.sql.SparkSession(sc)\n",
    "\n",
    "# hl.init(sc=sc, default_reference=REFERENCE_GENOME, log=LOG_FILE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a6d94eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pval(x):\n",
    "    if x < 0.0001:\n",
    "        return \"****\"\n",
    "    if x < 0.001:\n",
    "        return \"***\"\n",
    "    if x < 0.01:\n",
    "        return \"**\"\n",
    "    if x < 0.05:\n",
    "        return \"*\"\n",
    "    if x > 0.05:\n",
    "        return \"\"\n",
    "\n",
    "\n",
    "def pheno_search(x):\n",
    "    try:\n",
    "        return ukb_coding[ukb_coding.FieldID == int(x)].Field.values[0]\n",
    "    except:\n",
    "        # coding from https://biobank.ndph.ox.ac.uk/showcase/coding.cgi?id=100605\n",
    "        if x == \"6150_1\":\n",
    "            return \"Heart attack diagnosed by doctor\"\n",
    "        elif x == \"6150_2\":\n",
    "            return \"Angina diagnosed by doctor\"\n",
    "        elif x == \"6150_3\":\n",
    "            return \"Stroke diagnosed by doctor\"\n",
    "        elif x == \"6150_4\":\n",
    "            return \"High blood pressure diagnosed by doctor\"\n",
    "        elif x == \"myo\":\n",
    "            return \"myocardial disease incl. angina pectoris\"\n",
    "        elif x == \"3005\":\n",
    "            return \"Fracture resulting from simple fall\"\n",
    "        else:\n",
    "            return x\n",
    "\n",
    "\n",
    "def fix_df(df, TRAIT):\n",
    "    if TRAIT == \"BT\":\n",
    "        df[\"OR\"] = np.exp(df.BETA)\n",
    "        df[\"OR_upper\"] = np.exp(df.BETA + df.SE)\n",
    "        df[\"OR_lower\"] = np.exp(df.BETA - df.SE)\n",
    "        df[\"OR_se\"] = df[\"OR_upper\"] - df[\"OR_lower\"]\n",
    "        df[\"OR_se_l\"] = df[\"OR\"] - df[\"OR_lower\"]\n",
    "        df[\"OR_se_u\"] = df[\"OR_upper\"] - df[\"OR\"]\n",
    "    elif TRAIT == \"QT\":\n",
    "        df.loc[:, \"OR\"] = df.BETA\n",
    "        df.loc[:, \"OR_se\"] = df.SE\n",
    "\n",
    "    df.loc[:, \"Phenotype\"] = df.PHENO.apply(lambda x: pheno_search(x))\n",
    "    df.loc[:, \"pval\"] = np.power(10, -df[\"LOG10P\"])\n",
    "    df.loc[:, \"pval_e\"] = df.pval.apply(lambda x: f\"{x:.2f}\")\n",
    "    df.loc[:, \"pval_stars\"] = df[\"pval\"].apply(lambda x: pval(x))\n",
    "    df.loc[:, \"N_pos\"] = (2 * df[\"N\"] * df[\"A1FREQ\"]).astype(int)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36933c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [\n",
    "    f\"file:/mnt/project/data/results/{GENE}/{file}\"\n",
    "    for file in os.listdir(f\"/mnt/project/data/results/{GENE}\")\n",
    "]\n",
    "\n",
    "ukb_coding = pd.read_csv(\n",
    "    \"/opt/notebooks/gogoGPCR/data/misc/Data_Dictionary_Showcase.csv\",\n",
    "    error_bad_lines=False,\n",
    "    warn_bad_lines=False,\n",
    "    quotechar='\"',\n",
    ")\n",
    "\n",
    "custom_coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f2fd8cd",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "691924a9",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": [
    "recode = {\n",
    "    \"systolic\": \"Systolic blood pressure\",\n",
    "    \"diastolic\": \"Diastolic blood pressure\",\n",
    "    \"diabetes\": \"Type 2 diabetes\",\n",
    "    \"obesity\": \"Obesity\",\n",
    "    \"blood_pressure_med\": \"Blood pressure medication\",\n",
    "    \"cholesterol_med\": \"Cholesterol medication\",\n",
    "}\n",
    "\n",
    "df = pd.read_csv(files[0], delimiter=\" \", header=\"infer\", comment=\"#\").assign(\n",
    "    SOURCE=os.path.basename(files[0])\n",
    ")\n",
    "df = pd.concat(\n",
    "    [df]\n",
    "    + [\n",
    "        pd.read_csv(fp, delimiter=\" \", comment=\"#\").assign(SOURCE=os.path.basename(fp))\n",
    "        for fp in files[1:]\n",
    "    ],\n",
    "    axis=0,\n",
    ")\n",
    "\n",
    "df.loc[:, \"GENE\"] = df.ID.apply(lambda x: x.split(\".\")[0])\n",
    "df.loc[:, \"MASK\"] = df.ALLELE1.apply(lambda x: x.split(\".\", maxsplit=2)[0])\n",
    "df.loc[:, \"AAF\"] = df.ALLELE1.apply(lambda x: x.split(\".\", maxsplit=1)[-1])\n",
    "df.loc[:, \"FILE\"] = df.SOURCE.apply(\n",
    "    lambda x: x.split(\"/\")[-1].split(\"\\.\")[0].split(\"_\", maxsplit=4)[2:]\n",
    ")\n",
    "df.loc[:, \"TRAIT\"] = df.FILE.apply(lambda x: x[0])\n",
    "df.loc[:, \"PHENO\"] = df.FILE.apply(lambda x: x[-1].split(\".\")[0])\n",
    "df = df.drop([\"ID\", \"ALLELE0\", \"ALLELE1\", \"EXTRA\", \"SOURCE\", \"FILE\", \"TEST\"], axis=1)\n",
    "df.loc[:, \"PHENO\"] = df.loc[:, \"PHENO\"].apply(lambda x: recode.get(x, x))\n",
    "df = df.loc[df[\"AAF\"] != \"singleton\", :]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e5db654",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Set trait specific\n",
    "TRAIT = \"QT\"\n",
    "\n",
    "titles = {\n",
    "    \"BT\": \"Binary Traits\",\n",
    "    \"QT\": \"Quantitative Traits\",\n",
    "}\n",
    "title = titles[TRAIT]\n",
    "\n",
    "xlabs = {\"BT\": \"OR\", \"QT\": \"β\"}\n",
    "xlab = xlabs[TRAIT]\n",
    "\n",
    "masks = df.MASK.unique()\n",
    "\n",
    "# masks = [unique_masks[0], unique_masks[3], unique_masks[1]]\n",
    "\n",
    "interesting_traits = {\n",
    "    \"QT\": [\n",
    "        \"Diastolic blood pressure\",\n",
    "        \"Systolic blood pressure\",\n",
    "        \"Body mass index (BMI)\",\n",
    "        \"Body fat percentage\",\n",
    "    ],\n",
    "    \"BT\": [\n",
    "        \"Type 2 diabetes\",\n",
    "        \"Obesity\",\n",
    "    ],\n",
    "}\n",
    "interesting = interesting_traits[TRAIT]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dfd8674",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc7104f9",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Plotting options\n",
    "\n",
    "lw = 1.3\n",
    "ms = 10\n",
    "height = 2 * len(interesting)\n",
    "\n",
    "legend_elements = [\n",
    "    Line2D([0], [0], color=\"tab:grey\", marker=\"o\", ms=ms, lw=0, label=\"WT-like\"),\n",
    "    Line2D(\n",
    "        [0], [0], color=\"tab:orange\", marker=\"o\", ms=ms, lw=0, label=\"cAMP LOF only\"\n",
    "    ),\n",
    "    Line2D([0], [0], color=\"tab:purple\", marker=\"o\", ms=ms, lw=0, label=\"Both LOF\"),\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c4d4f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df.loc[df.TRAIT.eq(TRAIT), :]\n",
    "df2 = fix_df(df2, TRAIT=TRAIT)\n",
    "df2 = df2.loc[df2.Phenotype.isin(interesting)]\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32715f3e",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "plt_df = (\n",
    "    df2.loc[(df2.GENE == GENE)]\n",
    "    .sort_values(by=[\"Phenotype\", \"AAF\"], ascending=[True, False])\n",
    "    .groupby([\"Phenotype\", \"MASK\"])\n",
    "    .first()\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "plt_df = plt_df.loc[\n",
    "    ~plt_df.Phenotype.isin(\n",
    "        [\n",
    "            \"sev_obesity\",\n",
    "            \"Date E66 first reported (obesity)\",\n",
    "            \"Date I10 first reported (essential (primary) hypertension)\",\n",
    "            \"myocardial disease incl. angina pectoris\",\n",
    "            \"Pulse rate, automated reading\",\n",
    "            \"Heart attack diagnosed by doctor\",\n",
    "            \"Angina diagnosed by doctor\",\n",
    "            \"Fracture resulting from simple fall\",\n",
    "            \"Stroke diagnosed by doctor\",\n",
    "        ]\n",
    "    )\n",
    "]\n",
    "\n",
    "group_by_mean = pd.DataFrame(\n",
    "    {\"mean\": plt_df.groupby([\"Phenotype\"]).agg(\"mean\")[\"OR\"]}\n",
    ").reset_index()\n",
    "\n",
    "group_by_mean = group_by_mean.sort_values(by=\"mean\", ascending=False).reset_index()\n",
    "\n",
    "sorter = list(group_by_mean[\"Phenotype\"])\n",
    "\n",
    "plt_df.loc[:, \"Phenotype\"] = plt_df.loc[:, \"Phenotype\"].astype(\"category\")\n",
    "plt_df.loc[:, \"Phenotype\"].cat.set_categories(sorter, inplace=True)\n",
    "\n",
    "plt_df = plt_df.sort_values(\n",
    "    by=[\"Phenotype\", \"MASK\"], ascending=[True, False]\n",
    ").reset_index(drop=True)\n",
    "\n",
    "phenotypes = plt_df.Phenotype.unique()\n",
    "\n",
    "fig, axes = plt.subplots(nrows=len(phenotypes), sharex=True, figsize=(4, height))\n",
    "\n",
    "for ax in range(0, len(phenotypes)):\n",
    "    temp = plt_df.loc[plt_df.Phenotype.eq(phenotypes[ax]), :]\n",
    "    temp = temp.loc[plt_df.MASK.isin(masks)]\n",
    "\n",
    "    temps = [temp[temp.MASK == mask] for mask in masks]\n",
    "\n",
    "    if TRAIT == \"BT\":\n",
    "        xerrs = [[temp[\"OR_se_l\"].values, temp[\"OR_se_u\"].values] for temp in temps]\n",
    "\n",
    "    elif TRAIT == \"QT\":\n",
    "        xerrs = [temp[\"OR_se\"] for temp in temps]\n",
    "\n",
    "    axes[ax].errorbar(\n",
    "        temps[0][\"OR\"],\n",
    "        temps[0].index,\n",
    "        alpha=0.99,\n",
    "        xerr=xerrs[0],\n",
    "        fmt=\"o\",\n",
    "        c=\"tab:grey\",\n",
    "        ecolor=\"black\",\n",
    "        ms=ms,\n",
    "        mew=0.0,\n",
    "        mec=\"black\",\n",
    "        elinewidth=lw,\n",
    "    )\n",
    "    axes[ax].errorbar(\n",
    "        temps[1][\"OR\"],\n",
    "        temps[1].index,\n",
    "        alpha=0.99,\n",
    "        xerr=xerrs[1],\n",
    "        fmt=\"o\",\n",
    "        c=\"tab:purple\",\n",
    "        ecolor=\"black\",\n",
    "        ms=ms,\n",
    "        mew=0.0,\n",
    "        mec=\"black\",\n",
    "        elinewidth=lw,\n",
    "    )\n",
    "    # axes[ax].errorbar(\n",
    "    #    temps[2][\"OR\"],\n",
    "    #    temps[2].index,\n",
    "    #    alpha=0.99,\n",
    "    #    xerr=xerrs[2],\n",
    "    #    fmt=\"o\",\n",
    "    #    c=\"tab:orange\",\n",
    "    #    ecolor=\"black\",\n",
    "    #    ms=ms,\n",
    "    #    mew=0.0,\n",
    "    #    mec=\"black\",\n",
    "    #    elinewidth=lw,\n",
    "    # )\n",
    "\n",
    "    ax0 = axes[ax].twinx()\n",
    "    ax0.set_ylim([0.25, 3.25])\n",
    "    ax0.set_yticks(\n",
    "        [\n",
    "            1,\n",
    "            2,\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    y2labels = (\n",
    "        temp.N_pos.astype(str)\n",
    "        #    + temp.or_e.astype(str).str.ljust(5).values\n",
    "        #    + [\"(\"]\n",
    "        #    + temp.lower_e.values\n",
    "        #    + [\",\"]\n",
    "        #    + temp.upper_e.values\n",
    "        + [\"   p = \"]\n",
    "        + temp.pval_e.values\n",
    "    )\n",
    "\n",
    "    ax0.set_yticklabels(y2labels, fontsize=9, fontdict={\"family\": \"monospace\"})\n",
    "    ax0.tick_params(right=False)\n",
    "    ax0.spines[\"top\"].set_alpha(0)\n",
    "    ax0.spines[\"left\"].set_alpha(0)\n",
    "    ax0.spines[\"right\"].set_alpha(0)\n",
    "    ax0.spines[\"bottom\"].set_alpha(0)\n",
    "    ax0.grid(False)\n",
    "    # axes[ax].invert_xaxis()\n",
    "    # only show every 3rd yticklabel\n",
    "    labels = [l if i % 3 == 0 else \"\" for i, l in enumerate(temp.Phenotype)]\n",
    "    axes[ax].set(yticks=temp.index, yticklabels=labels[::-1])\n",
    "\n",
    "    # axes[ax].axvline(x=0, linestyle=\"--\", color=\"#4f4f4f\")\n",
    "    axes[ax].tick_params(left=False)\n",
    "\n",
    "    if TRAIT == \"BT\":\n",
    "        axes[ax].set_xlim([-0.2, 5.2])\n",
    "        axes[ax].axvline(x=1, linestyle=\":\", color=\"#4f4f4f\")\n",
    "    elif TRAIT == \"QT\":\n",
    "        axes[ax].set_xlim([-0.22, 0.82])\n",
    "        axes[ax].axvline(x=0, linestyle=\":\", color=\"#4f4f4f\")\n",
    "\n",
    "    axes[ax].spines[\"top\"].set_alpha(0)\n",
    "    axes[ax].spines[\"left\"].set_alpha(0)\n",
    "    axes[ax].spines[\"right\"].set_alpha(0)\n",
    "    if ax != len(phenotypes) - 1:\n",
    "        axes[ax].spines[\"bottom\"].set_alpha(0)\n",
    "        axes[ax].tick_params(bottom=False)\n",
    "\n",
    "    if ax == len(phenotypes) - 1:\n",
    "        axes[ax].set_xlabel(xlab)\n",
    "        # ax0.legend(handles = legend_elements, loc = \"lower right\")\n",
    "#\n",
    "# axes[len(phenotypes) - 1].set_xticks([0,1,2,3,4,5,6,7,])\n",
    "for ax in axes.flat:\n",
    "    ax.margins(0.3)\n",
    "\n",
    "plt.subplots_adjust(right=1)\n",
    "# fig.suptitle(title)\n",
    "\n",
    "# plt.savefig(f\"/opt/notebooks/gogoGPCR/tmp/{GENE}_{TRAIT}_new.svg\", dpi = 600, bbox_inches=\"tight\", format = \"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eef3c8a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt_df[\"N_pos\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c60db69",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"N_alleles\"] = df[\"N\"] * df[\"A1FREQ\"] * 2\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "86ccb1d9",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "pheno_search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4ccca43",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "encoding": "# -*- coding: utf-8 -*-",
   "formats": "ipynb,py:percent"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
